partition_table.go

```diff
@@ -194,43 +194,58 @@ func (p *PartitionTable) Close() error {
 func (p *PartitionTable) createStorage(ctx context.Context) (*storageProxy, error) {
 	var (
 		err  error
+		errs = new(multierr.Errors)
 		st   storage.Storage
 		done = make(chan struct{})
 	)
+
 	start := time.Now()
 	ticker := time.NewTicker(5 * time.Minute)
 	defer ticker.Stop()
+	defer close(done)
 	go func() {
-		defer close(done)
-		st, err = p.builder(p.topic, p.partition)
-		if err != nil {
-			return
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-done:
+				return
+			case <-ticker.C:
+				p.log.Printf("creating storage for topic %s/%d for %.1f minutes ...", p.topic, p.partition, time.Since(start).Minutes())
+			}
 		}
-		err = st.Open()
 	}()
 
-WaitLoop:
-	for {
-		select {
-		case <-ctx.Done():
-			return nil, nil
-		case <-ticker.C:
-			p.log.Printf("creating storage for topic %s/%d for %.1f minutes ...", p.topic, p.partition, time.Since(start).Minutes())
-		case <-done:
-			p.log.Debugf("finished building storage for topic %s/%d in %.1f minutes", p.topic, p.partition, time.Since(start).Minutes())
-			if err != nil {
-				return nil, fmt.Errorf("error building storage: %v", err)
-			}
-			break WaitLoop
+	st, err = p.builder(p.topic, p.partition)
+	if err != nil {
+		return nil, fmt.Errorf("error building storage: %v", err)
+	}
+	err = st.Open()
+	if err != nil {
+		errs.Collect(st.Close())
+		errs.Collect(fmt.Errorf("error opening storage: %v", err))
+		return nil, errs.NilOrError()
+	}
+
+	// close the db if context was cancelled before the builder returned
+	select {
+	case <-ctx.Done():
+		err = st.Close()
+		// do not report an error for only cancelling the context
```

Comment:
the comment is a bit misleading, as it says we're not reporting the error, but actually the errors are added to the collector?

Reply:
We only report the context error if `st.Close()` returned error. Otherwise, no error is reported even if context had one.
I'll rephrase to: `only collect context error if Close() errored out`

Reply:
thanks, much easier to understand